lrpar/src/lib/parser.rs

```diff
@@ -91,11 +101,18 @@ pub struct Parser<'a, 'b: 'a, 'input: 'b, StorageT: 'static + Eq + Hash, ActionT
     // In the long term, we should remove the `lexemes` field entirely, as the `NonStreamingLexer` API is
     // powerful enough to allow us to incrementally obtain lexemes and buffer them when necessary.
     pub(crate) lexemes: Vec<Lexeme<StorageT>>,
-    actions: &'a [ActionFn<'a, 'b, 'input, StorageT, ActionT>],
+    actions: &'a [ActionFn<'a, 'b, 'input, 'arg, StorageT, ActionT, ParamT>],
```

Comment:
This might be a silly question but I've just realised that I'm not sure why an action can *return* a `ParamT`. Is this something to do with ownership problems?

Reply:
Not a silly question at all, it strikes quite to the heart of it, If we look at the loop in `parser::Parser::lr`, and `lr_upto`, where we call the action inside parser (because param isn't `Copy` bounded, it moves the param field, so we have to return it there, in order to regain ownership.

I'll add some comments at the places, but I'm not sure how to do so in this thread.

Reply:
If we're doing this sort of moving, it suggests to me that we really intend to pass the param as a reference? Or is this because we want to allow mutation?

Reply:
Yeah its because of mutation, we have to move the mutable reference around to enforce single-ownership.

But it does bring up a point that I haven't tested this with a non-reference param.

Reply:
Maybe passing it as `&mut` would do the trick?

Reply:
Not quite certain, I can give it a try though, I had difficulty trying to use the fancy types (higher ranked lifetime bounds) here, which allows us to show that the `&mut` isn't captured by the `Fn` closure.  So this was kind of a last resort, but perhaps starting from a place where it now works I can figure that out.

Reply:
Fingers crossed!

Reply:
Well, i'm sorry to say that I don't think it is possible, the problem is that `ActionFn` takes an owned `vec::Drain`, the HRTB works via saying `for <'c>  Fn(...)`, essentially saying the parameters can live either less than or greater than the lifetime of the closure (Essentially a `Fn` equivalent of a `fn`, but the closure itself owns the `vec::Drain`, and so must have a self which captures it.  There isn't any fine-grained per parameter way to type that I believe.

Reply:
I think I sort-of understand. I guess the challenge with passing in `ParamT` and receiving back a `ParamT` is that we've semi-destroyed the relationship between the two from the type system's perspective. This might be a very dumb thing to say, but can we cheat by wrapping the parameter in something (maybe, at its heart, an `UnsafeCell`)? I might be clutching at straws here...

Reply:
I'm really not familiar with `unsafe` in general, but I *guess* you can do anything, it would add some difficulty in getting stuff out of the unsafe cell in the way the user-code is expecting the environment.

One thing I was wondering, is whether it is possible to use a thread_local for either the `Drain` or the `ParamT`, If we can just get one of those out of the `Fn` signature while somehow avoiding moving `ParamT`.

Reply:
We definitely can't do that for the `Drain`. *But* it might be possible to convert the `Drain` into a slice which would be passed by reference -- would that make any difference?

Reply:
I think that it would be possible to get a reference to a slice to work with the `for<'c>` quantifier yes.  I'll try and work on a more reflective example in the playground again.